Optimize HSET by skipping listpack element count recalculation when no new field is added

vitahlin opened the pull request:
## Background

The HSET command is used to insert or update fields in a hash. When the hash is encoded using the listpack encoding, the number of elements in the listpack is recalculated each time the HSET operation is executed. 

This is the method used to calculate the number of elements in a listpack:
https://github.com/redis/redis/blob/6ec78b47427e3d5b7dcda065f7d405bbbb47a5a4/src/listpack.c#L532-L554

When the number of elements in the listpack is less than 65,536, the overhead is minimal because the cached length can be accessed directly. However, when the number exceeds 65,536, the entire listpack must be traversed to calculate the length, which introduces significant overhead.


## Optimization

When executing the HSET command, if no new fields are added (i.e., no new key-value pairs are inserted), the `hashTypeLength` method is not called, thereby skipping the recalculation of the listpack element count.

## Test

### the element count of listpack less then 65536

Test command:
```shell
memtier_benchmark \
  --pipeline 1 \
  --command "HSET test_hash f1 v1 f2 v2 f3 v3 f4 v4 f5 v5 f6 v6 f7 v7 f8 v8 f9 v9 f10 v10 f11 v11 f12 v12 f13 v13 f14 v14 f15 v15 f16 v16 f17 v17 f18 v18 f19 v19 f20 v20" \
  --test-time 60 \
  -c 1 \
  -t 1 \
  --hide-histogram
```

Before commit:
```shell
ALL STATS
==================================================================================================
Type         Ops/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec 
--------------------------------------------------------------------------------------------------
Hsets       35262.37         0.02890         0.03100         0.06300         0.09500     17080.21 
Totals      35262.37         0.02890         0.03100         0.06300         0.09500     34160.43
```

After commit:
```shell
ALL STATS
==================================================================================================
Type         Ops/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec 
--------------------------------------------------------------------------------------------------
Hsets       35434.14         0.02877         0.03100         0.05500         0.09500     17163.41 
Totals      35434.14         0.02877         0.03100         0.05500         0.09500     34326.83
```

Here is one of my test results. When the count of element(20) is less than 65536, the two test result are quite similar, and the optimization effect is not significant.


### The element count of listpack exceeds 65,536 (reaching 66,000). 


Test command:

1. Modify the default value of `hash-max-listpack-entries` so that the maximum length of a listpack can exceed 65536, and the listpack will not be converted to a hash type”:
```shell
$ redis-cli CONFIG SET hash-max-listpack-entries 70000
```

2. Use a shell script to insert 66,000 entries into the listpack:
```shell
#!/bin/bash

BATCH_SIZE=100
TOTAL_FIELDS=66000
KEY="test_hash"
REDIS_CLI="redis-cli"

$REDIS_CLI DEL $KEY > /dev/null

i=1
while [ $i -le $TOTAL_FIELDS ]; do
  CMD="HSET $KEY"

  for ((j=0; j<$BATCH_SIZE && $i<=$TOTAL_FIELDS; j+=2)); do
    CMD+=" f$i v$i"
    i=$((i+1))
  done

  $REDIS_CLI $CMD
done

echo "Inserted $TOTAL_FIELDS fields into $KEY."
```

```shell
memtier_benchmark \
  --pipeline 1 \
  --command "HSET test_hash f1 v1 f2 v2 f3 v3 f4 v4 f5 v5 f6 v6 f7 v7 f8 v8 f9 v9 f10 v10 f11 v11 f12 v12 f13 v13 f14 v14 f15 v15 f16 v16 f17 v17 f18 v18 f19 v19 f20 v20" \
  --test-time 60 \
  -c 1 \
  -t 1 \
  --hide-histogram
```


Before commit:
```shell
ALL STATS
==================================================================================================
Type         Ops/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec 
--------------------------------------------------------------------------------------------------
Hsets          75.49        13.24575        13.18300        14.07900        21.11900        27.72 
Totals         75.49        13.24575        13.18300        14.07900        21.11900        55.44
```

After commit:
```shell
ALL STATS
==================================================================================================
Type         Ops/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec 
--------------------------------------------------------------------------------------------------
Hsets        1384.68         0.72210         0.72700         0.80700         0.99900       508.44 
Totals       1384.68         0.72210         0.72700         0.80700         0.99900      1016.87

```

vitahlin commented:
The CI failed, and there are also some cases I didn’t consider. I’m currently investigating.

sundb commented on src/t_hash.c:
```diff
@@ -929,8 +929,10 @@ int hashTypeSet(redisDb *db, robj *o, sds field, sds value, int flags) {
+        /* Check if the listpack needs to be converted to a hash table.
+         * We do not subtract expired field here, if there is no new field inserted,
+         * we also skip recomputing the listpack element count, avoiding unnecessary overhead. */
+        if (!update && hashTypeLength(o, 0) > server.hash_max_listpack_entries)
```
but we still might replace a small field with a large field.
in this case, does hash still need to be converted to ht?

vitahlin replied:
The size of the field has been validated here:
https://github.com/redis/redis/blob/ba88a7fbb6ec7cb7aeda61e1981146e1a21dee43/src/t_hash.c#L898-L902

The function `hashTypeLength` only checks the number of elements in the hash.

sundb commented:
> The test cases `Module scan hash dict` and `Module scan hash listpack with hexpire` use the same key `hh`. As a result, the `Module scan hash dict` test fails because the key `hh` already contains the same field, so the `hmset f3 v3` operation does not trigger an insertion. From what I’ve seen in other tests, when the same key is used, it’s typically deleted or reset to avoid side effects between tests. Maybe we should adjust these tests to make them independent of each other?

yes, we should do that.

fcostaoliveira commented:
@sundb @vitahlin i've added a testcase to cover the listpack hash usage that matches the above optimization (didn't changed the default listpack limits given we want to optimize towards the defaults), BUT was not able to check the boost mentioned in the PR comment even with 500 fields on each hash. 

|                                                                                                                                              Test Case                                                                                                                                              |Baseline /redis unstable (median obs. +- std.dev)|Comparison vitahlin/redis a2dfe4ea127026aea244e6f61120f6c7941831b6 (median obs. +- std.dev)|% change (lower-better)|  Note   |
|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------:|------------------------------------------------------------------------------------------:|-----------------------|---------|
|[memtier_benchmark-1Kkeys-hash-listpack-500-fields-update-20-fields-with-1B-to-64B-values](https://github.com/redis/redis-benchmarks-specification/blob/main/redis_benchmarks_specification/test-suites/memtier_benchmark-1Kkeys-hash-listpack-500-fields-update-20-fields-with-1B-to-64B-values.yml)|                                              2.4|                                                                                        2.4|1.3%                   |No Change|

I'll trigger the full HASH group benchmarks to be sure. 
In the meantime @vitahlin can you merge unstable into it? want to avoid any diffs due to that.

kaplanben commented:
![Logo](https://cdn.ast.checkmarx.net/integrations/logo/Checkmarx.png)
**Checkmarx One – Scan Summary & Details** – [df147abd\-f9c3\-45fd\-9286\-cb148e987003](https://gov-il.ast.checkmarx.net/projects/6e34a6db-7b74-4c41-8bc6-d52b8f2a3534/scans?id=df147abd-f9c3-45fd-9286-cb148e987003&branch=vitahlin%2Foptimize-hset-listpack-conversion)
<!--new issues section--><details><summary><b>New Issues (7)</b></summary>
Checkmarx found the following issues in this Pull Request

Severity|Issue|Source File / Package|Checkmarx Insight
---|---|---|---
![CRITICAL](https://cdn.ast.checkmarx.net/integrations/severity/Critical.png)|[Buffer_Overflow_Wrong_Buffer_Size](https://gov-il.ast.checkmarx.net/results/6e34a6db-7b74-4c41-8bc6-d52b8f2a3534/df147abd-f9c3-45fd-9286-cb148e987003/sast/description/131/9017219199198855303)|/src/redis\-cli\.c: [3677](https://github.com/vitahlin/redis/blob/optimize-hset-listpack-conversion//src/redis-cli.c#L3677)|<details><summary>details</summary>The buffer buf created in /src/redis\-cli\.c at line 3677 is written to a buffer in /deps/hiredis/sds\.c at line 234 by newsh, but an error in calc...<br>`ID: %2BpSSxZAM7xfUiads1egmyYebO5I%3D`</details>[Attack Vector](https://gov-il.ast.checkmarx.net/results/df147abd-f9c3-45fd-9286-cb148e987003/6e34a6db-7b74-4c41-8bc6-d52b8f2a3534/sast?result-id=%2BpSSxZAM7xfUiads1egmyYebO5I%3D&redirect=true)
![CRITICAL](https://cdn.ast.checkmarx.net/integrations/severity/Critical.png)|[Buffer_Overflow_Wrong_Buffer_Size](https://gov-il.ast.checkmarx.net/results/6e34a6db-7b74-4c41-8bc6-d52b8f2a3534/df147abd-f9c3-45fd-9286-cb148e987003/sast/description/131/9017219199198855303)|/src/redis\-cli\.c: [3677](https://github.com/vitahlin/redis/blob/optimize-hset-listpack-conversion//src/redis-cli.c#L3677)|<details><summary>details</summary>The buffer buf created in /src/redis\-cli\.c at line 3677 is written to a buffer in /deps/hiredis/sds\.c at line 234 by hdrlen, but an error in cal...<br>`ID: zN%2FI3F1XTVrKpHuopU6EZZmWXt4%3D`</details>[Attack Vector](https://gov-il.ast.checkmarx.net/results/df147abd-f9c3-45fd-9286-cb148e987003/6e34a6db-7b74-4c41-8bc6-d52b8f2a3534/sast?result-id=zN%2FI3F1XTVrKpHuopU6EZZmWXt4%3D&redirect=true)
![CRITICAL](https://cdn.ast.checkmarx.net/integrations/severity/Critical.png)|[Buffer_Overflow_Wrong_Buffer_Size](https://gov-il.ast.checkmarx.net/results/6e34a6db-7b74-4c41-8bc6-d52b8f2a3534/df147abd-f9c3-45fd-9286-cb148e987003/sast/description/131/9017219199198855303)|/deps/linenoise/linenoise\.c: [1200](https://github.com/vitahlin/redis/blob/optimize-hset-listpack-conversion//deps/linenoise/linenoise.c#L1200)|<details><summary>details</summary>The buffer buf created in /deps/linenoise/linenoise\.c at line 1200 is written to a buffer in /deps/hiredis/sds\.c at line 97 by sh, but an error i...<br>`ID: oykVSjUcVC%2FEMplDwW4P3YG7%2FzE%3D`</details>[Attack Vector](https://gov-il.ast.checkmarx.net/results/df147abd-f9c3-45fd-9286-cb148e987003/6e34a6db-7b74-4c41-8bc6-d52b8f2a3534/sast?result-id=oykVSjUcVC%2FEMplDwW4P3YG7%2FzE%3D&redirect=true)
![CRITICAL](https://cdn.ast.checkmarx.net/integrations/severity/Critical.png)|[Buffer_Overflow_Wrong_Buffer_Size](https://gov-il.ast.checkmarx.net/results/6e34a6db-7b74-4c41-8bc6-d52b8f2a3534/df147abd-f9c3-45fd-9286-cb148e987003/sast/description/131/9017219199198855303)|/src/redis\-cli\.c: [10594](https://github.com/vitahlin/redis/blob/optimize-hset-listpack-conversion//src/redis-cli.c#L10594)|<details><summary>details</summary>The buffer argv created in /src/redis\-cli\.c at line 10594 is written to a buffer in /deps/hiredis/sds\.c at line 97 by sh, but an error in calcul...<br>`ID: eStOv%2FTaWfWWBCJCCgzT7mgYJU0%3D`</details>[Attack Vector](https://gov-il.ast.checkmarx.net/results/df147abd-f9c3-45fd-9286-cb148e987003/6e34a6db-7b74-4c41-8bc6-d52b8f2a3534/sast?result-id=eStOv%2FTaWfWWBCJCCgzT7mgYJU0%3D&redirect=true)
![CRITICAL](https://cdn.ast.checkmarx.net/integrations/severity/Critical.png)|[Buffer_Overflow_Wrong_Buffer_Size](https://gov-il.ast.checkmarx.net/results/6e34a6db-7b74-4c41-8bc6-d52b8f2a3534/df147abd-f9c3-45fd-9286-cb148e987003/sast/description/131/9017219199198855303)|/deps/linenoise/linenoise\.c: [1166](https://github.com/vitahlin/redis/blob/optimize-hset-listpack-conversion//deps/linenoise/linenoise.c#L1166)|<details><summary>details</summary>The buffer fgetc created in /deps/linenoise/linenoise\.c at line 1166 is written to a buffer in /deps/hiredis/sds\.c at line 97 by sh, but an error...<br>`ID: v3h9G7I8PLSutWNyC8k4gGzAdDA%3D`</details>[Attack Vector](https://gov-il.ast.checkmarx.net/results/df147abd-f9c3-45fd-9286-cb148e987003/6e34a6db-7b74-4c41-8bc6-d52b8f2a3534/sast?result-id=v3h9G7I8PLSutWNyC8k4gGzAdDA%3D&redirect=true)
![MEDIUM](https://cdn.ast.checkmarx.net/integrations/severity/Medium.png)|[Divide_By_Zero](https://gov-il.ast.checkmarx.net/results/6e34a6db-7b74-4c41-8bc6-d52b8f2a3534/df147abd-f9c3-45fd-9286-cb148e987003/sast/description/369/10360811894101193280)|/modules/vector\-sets/fastjson\_test\.c: [121](https://github.com/vitahlin/redis/blob/optimize-hset-listpack-conversion//modules/vector-sets/fastjson_test.c#L121)|<details><summary>details</summary>The application performs an illegal operation in generate\_random\_string, in /modules/vector\-sets/fastjson\_test\.c\. In line 121, the program at...<br>`ID: qiowoZ%2FDUFf8wA3ZCvKY8M0GHks%3D`</details>[Attack Vector](https://gov-il.ast.checkmarx.net/results/df147abd-f9c3-45fd-9286-cb148e987003/6e34a6db-7b74-4c41-8bc6-d52b8f2a3534/sast?result-id=qiowoZ%2FDUFf8wA3ZCvKY8M0GHks%3D&redirect=true)
![MEDIUM](https://cdn.ast.checkmarx.net/integrations/severity/Medium.png)|[Divide_By_Zero](https://gov-il.ast.checkmarx.net/results/6e34a6db-7b74-4c41-8bc6-d52b8f2a3534/df147abd-f9c3-45fd-9286-cb148e987003/sast/description/369/10360811894101193280)|/src/redis\-cli\.c: [6040](https://github.com/vitahlin/redis/blob/optimize-hset-listpack-conversion//src/redis-cli.c#L6040)|<details><summary>details</summary>The application performs an illegal operation in clusterManagerNodeMasterRandom, in /src/redis\-cli\.c\. In line 6053, the program attempts to divi...<br>`ID: Wdmj3BiFZXbdNClmOY%2Fr1waYywk%3D`</details>[Attack Vector](https://gov-il.ast.checkmarx.net/results/df147abd-f9c3-45fd-9286-cb148e987003/6e34a6db-7b74-4c41-8bc6-d52b8f2a3534/sast?result-id=Wdmj3BiFZXbdNClmOY%2Fr1waYywk%3D&redirect=true)

</details>

<!--new issues section--><!--fix issues section--><details><summary><b>Fixed Issues (5)</b></summary>
Great job! The following issues were fixed in this Pull Request

Severity|Issue|Source File / Package
---|---|---
![CRITICAL](https://cdn.ast.checkmarx.net/integrations/severity/Critical.png)|~~[Buffer_Overflow_Wrong_Buffer_Size](https://gov-il.ast.checkmarx.net/results/6e34a6db-7b74-4c41-8bc6-d52b8f2a3534/df147abd-f9c3-45fd-9286-cb148e987003/sast/description/131/9017219199198855303)~~|/src/redis\-cli\.c: [3677](https://github.com/vitahlin/redis/blob/optimize-hset-listpack-conversion//src/redis-cli.c#L3677)
![CRITICAL](https://cdn.ast.checkmarx.net/integrations/severity/Critical.png)|~~[Buffer_Overflow_Wrong_Buffer_Size](https://gov-il.ast.checkmarx.net/results/6e34a6db-7b74-4c41-8bc6-d52b8f2a3534/df147abd-f9c3-45fd-9286-cb148e987003/sast/description/131/9017219199198855303)~~|/src/redis\-cli\.c: [3677](https://github.com/vitahlin/redis/blob/optimize-hset-listpack-conversion//src/redis-cli.c#L3677)
![CRITICAL](https://cdn.ast.checkmarx.net/integrations/severity/Critical.png)|~~[Buffer_Overflow_Wrong_Buffer_Size](https://gov-il.ast.checkmarx.net/results/6e34a6db-7b74-4c41-8bc6-d52b8f2a3534/df147abd-f9c3-45fd-9286-cb148e987003/sast/description/131/9017219199198855303)~~|/src/redis\-cli\.c: [3677](https://github.com/vitahlin/redis/blob/optimize-hset-listpack-conversion//src/redis-cli.c#L3677)
![CRITICAL](https://cdn.ast.checkmarx.net/integrations/severity/Critical.png)|~~[Buffer_Overflow_Wrong_Buffer_Size](https://gov-il.ast.checkmarx.net/results/6e34a6db-7b74-4c41-8bc6-d52b8f2a3534/df147abd-f9c3-45fd-9286-cb148e987003/sast/description/131/9017219199198855303)~~|/src/redis\-cli\.c: [3677](https://github.com/vitahlin/redis/blob/optimize-hset-listpack-conversion//src/redis-cli.c#L3677)
![MEDIUM](https://cdn.ast.checkmarx.net/integrations/severity/Medium.png)|~~[Divide_By_Zero](https://gov-il.ast.checkmarx.net/results/6e34a6db-7b74-4c41-8bc6-d52b8f2a3534/df147abd-f9c3-45fd-9286-cb148e987003/sast/description/369/10360811894101193280)~~|/deps/jemalloc/src/nstime\.c: [149](https://github.com/vitahlin/redis/blob/optimize-hset-listpack-conversion//deps/jemalloc/src/nstime.c#L149)
</details>

<!--fix issues section-->

vitahlin commented:
Thanks for reviewing this PR!

In this PR, I updated hashTypeSet and adjusted some related tests. After review, I realized the optimization only helps in extreme cases (listpacks > 65,536 entries) and has very limited practical impact. The test updates were minor and mainly avoid key reuse side effects.

Given this, I think it’s better to close this PR for now and revisit any test changes when we touch the related hash code in the future.
